core/src/main/scala/scalaz/Name.scala

```diff
@@ -52,13 +52,31 @@ object Name {
   }
 }
 
-object Need {
+/** Synchronized, thread-safe version of `Need` */
+object ThreadSafeNeed {
   def apply[A](a: => A): Need[A] = {
     new Need[A] {
-      private[this] lazy val value0: A = a
-      def value = value0
+      lazy val value: A = a
+    }
+  }
+}
+
+object Need {
```

Comment:
Changed the implementation of `Need` to not use a lazy val. `ThreadSafeNeed` uses a `lazy val` to synchronize the memo.